metasrv/src/watcher/watcher_manager.rs

```diff
@@ -0,0 +1,242 @@
+            subscriber: WatcherStateMachineSubscriber { event_tx },
+        }
+    }
+
+    pub fn stop(&self) {
+        let _ = self.event_tx.send(WatcherEvent::ShutdownEvent(()));
+    }
+
+    pub fn create_watcher_stream(&self, request: WatchRequest, tx: WatcherStreamSender) {
+        let create: CreateWatcherEvent = (request, tx);
+        let _ = self.event_tx.send(WatcherEvent::CreateWatcherEvent(create));
+    }
+}
+
+impl WatcherManagerCore {
+    #[tracing::instrument(level = "trace", skip(self))]
+    async fn watcher_manager_main(mut self) {
+        loop {
+            tokio::select! {
+                event = self.event_rx.recv() => {
+                    if let Some(event) = event {
+                        match event {
+                            WatcherEvent::CreateWatcherEvent((req, tx)) => {self.create_watcher_stream(req, tx).await;},
+                            WatcherEvent::StateMachineKvDataEvent(kv) => {self.notify_event(kv).await;},
+                            WatcherEvent::ShutdownEvent(_) => {
+                                tracing::info!("watcher manager has been shutdown");
+                                break;
+                            }
+                        }
+                    }
+                },
+            }
+        }
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    fn close_stream(&mut self, id: WatcherId) {
+        let watcher = self.watcher_streams.get(&id);
+        if let Some(watcher) = watcher {
+            let range = watcher.key.clone()..watcher.key_end.clone();
+            self.watcher_range_set.remove(range, WatcherKey {
+                id,
+                filter: FilterType::All,
+            });
+
+            self.watcher_streams.remove(&id);
+        }
+    }
+
+    async fn notify_event(&mut self, kv: StateMachineKvData) {
+        let set = self.watcher_range_set.get_by_point(&kv.key);
+        if set.is_empty() {
+            return;
+        }
+        let current = kv.current.as_ref().map(|current| current.data.clone());
+
+        let prev = kv.prev.as_ref().map(|prev| prev.data.clone());
+
+        let is_delete_event = kv.current.is_none();
+        let mut remove_watcher_id: Vec<WatcherId> = vec![];
+
+        for range_key in set.iter() {
+            let watcher_id = range_key.key.id;
+            let filter = range_key.key.filter;
+
+            // filter out event
+            if (filter == FilterType::Delete && !is_delete_event)
+                || (filter == FilterType::Update && is_delete_event)
+            {
+                continue;
+            }
+
+            if let Some(stream) = self.watcher_streams.get(&watcher_id) {
+                assert_eq!(stream.id, watcher_id);
+                let resp = WatchResponse {
+                    event: Some(Event {
+                        key: kv.key.clone(),
+                        current: current.clone(),
+                        prev: prev.clone(),
```

Comment:
If the `seq` is discarded, when the client receives an event, it may need to retrieve the `seq` with another RPC. Since several meta operations rely on `seq` to guarantee consistency.